Data/ByteString/Base64/Internal.hs

```diff
@@ -205,48 +206,104 @@ decodeWithTable padding decodeFP bs@(PS !fp !o !l) = unsafePerformIO $
       else io
 
-
     go (PS !sfp !soff !slen) = do
       dfp <- mallocByteString dlen
-      withForeignPtr decodeFP $ \ !decptr -> do
-        let finish dbytes = return . Right $! if dbytes > 0
-                                              then PS dfp 0 dbytes
-                                              else B.empty
-            bail = return . Left
-        withForeignPtr sfp $ \ !sptr -> do
-          let sEnd = sptr `plusPtr` (slen + soff)
-              look p = do
-                ix <- fromIntegral `fmap` peek8 p
-                v <- peek8 (decptr `plusPtr` ix)
-                return $! fromIntegral v :: IO Word32
-              fill !dp !sp !n
-                | sp >= sEnd = finish n
-                | otherwise = {-# SCC "decodeWithTable/fill" #-} do
-                a <- look sp
-                b <- look (sp `plusPtr` 1)
-                c <- look (sp `plusPtr` 2)
-                d <- look (sp `plusPtr` 3)
-                let w = (a `shiftL` 18) .|. (b `shiftL` 12) .|.
-                        (c `shiftL` 6) .|. d
-                if a == done || b == done
-                  then bail $ "invalid padding near offset " ++
-                       show (sp `minusPtr` sptr)
-                  else if a .|. b .|. c .|. d == x
-                  then bail $ "invalid base64 encoding near offset " ++
-                       show (sp `minusPtr` sptr)
-                  else do
-                    poke8 dp $ fromIntegral (w `shiftR` 16)
-                    if c == done
-                      then finish $ n + 1
-                      else do
-                        poke8 (dp `plusPtr` 1) $ fromIntegral (w `shiftR` 8)
-                        if d == done
-                          then finish $! n + 2
-                          else do
-                            poke8 (dp `plusPtr` 2) $ fromIntegral w
-                            fill (dp `plusPtr` 3) (sp `plusPtr` 4) (n+3)
-          withForeignPtr dfp $ \dptr ->
-            fill dptr (sptr `plusPtr` soff) 0
+      withForeignPtr decodeFP $ \ !decptr ->
+        withForeignPtr sfp $ \sptr ->
+        withForeignPtr dfp $ \dptr -> do
+          let !end = sptr `plusPtr` (slen + soff)
+          decodeLoop
+            decptr
+            (plusPtr sptr soff)
+            dptr
+            end
+            dfp
+
+decodeLoop
+    :: Ptr Word8
+    -> Ptr Word8
+    -> Ptr Word8
+    -> Ptr Word8
+    -> ForeignPtr Word8
+    -> IO (Either String ByteString)
+decodeLoop !dtable !sptr !dptr !end !dfp = go dptr sptr 0
+  where
+    err p = return . Left
+      $ "invalid character at offset: "
+      ++ show (p `minusPtr` sptr)
+
+    padErr p =  return . Left
+      $ "invalid padding at offset: "
+      ++ show (p `minusPtr` sptr)
+
+    look :: Ptr Word8 -> IO Word32
+    look !p = do
+      !i <- peekByteOff p 0 :: IO Word8
+      !v <- peekByteOff dtable (fromIntegral i) :: IO Word8
+      return (fromIntegral v)
+
+    go !dst !src !n
```

Comment:
do we really need to compute `n` on every iteration? can't we just recover `n` when we actually need it (afaics this is primarily for the `finish` call-site?) by subtracting the `dst` pointer from the `dptr`?

Reply:
I'm following the convention set by the library. I could certainly do this, sure. I find it easier to count, though, and it shouldn't have any tangible impact on performance

Reply:
Either way, done.